### PR TITLE
Setup gradle plugin for publishing

### PR DESCRIPTION
**Pull request by mdedetrich**
References: https://github.com/apache/incubator-pekko-grpc/issues/113

Only thing that is missing is signing (see https://github.com/gradle-nexus/publish-plugin#configure-signing) but lets do that separately.

### CODE REVIEW

File: gradle-plugin/build.gradle [hunk 1]
```diff
@@ -23,18 +23,43 @@ gradlePlugin {
...
 }
 
-pluginBundle {
-  website = 'https://pekko.apache.org/docs/pekko-grpc/current/'
-  vcsUrl = 'https://github.com/apache/incubator-pekko-grpc'
-  description = 'Building streaming gRPC servers and clients on top of Apache Pekko Streams'
-  tags = ['pekko', 'streams', 'reactive']
+publishing {
+    publications.withType(MavenPublication) {
+        pom {
```
Comment (mdedetrich) [inline gradle-plugin/build.gradle:28]: A lot of the pom details were outdated/missing. This data was retrieved by copying it from the `pom` of one of the sbt artifacts of this project

File: gradle-plugin/build.gradle [hunk 2]
```diff
@@ -23,18 +23,43 @@ gradlePlugin {
...
 
-  plugins {
-    pekkoGrpcPlugin {
-      id = 'org.apache.pekko.grpc.gradle'
-      displayName = 'Apache Pekko gRPC'
+nexusPublishing {
+    repositories {
+        sonatype {
+            nexusUrl.set(uri("https://repository.apache.org/"))
+            snapshotRepositoryUrl.set(uri("https://repository.apache.org/content/repositories/snapshots/"))
```
Comment (pjfanning) [inline gradle-plugin/build.gradle:58]: so when we need a release manager (a human) to do a release, they will need to hand edit this?

Comment (pjfanning) [reply gradle-plugin/build.gradle:58]: the release URL will be something like https://repository.apache.org/service/local/staging - I need to track the correct path but it is something similar to this.

Comment (mdedetrich) [reply gradle-plugin/build.gradle:58]: > so when we need a release manager (a human) to do a release, they will need to hand edit this?

Nope, aside from maybe the version (but we have the same issue with sbt)

> the release URL will be something like https://repository.apache.org/service/local/staging - I need to track the correct path but it is something similar to this.

This should already be handled by the various `publishToSonatype` tasks and its variants as described in https://github.com/gradle-nexus/publish-plugin#configure-signing. The `https://repository.apache.org/service/local/staging` uri is nothing that should ever be set, thats an internal detail of how bundle/standard publishing is done by the plugin. If you check sbt-apache-sonatype (https://github.com/mdedetrich/sbt-apache-sonatype/blob/main/src/main/scala/org/mdedetrich/apache/sonatype/ApacheSonatypePlugin.scala#L36) you can see we only set the URI in on place.

In summary its the role of the plugins (i.e. gradle nexus publish-plugin/sbt-sonatype) to configure the staging repos, you only need to set the base URI. What may be a difference with the gradle nexus publish-plugin is that it may not support bundle publishing, but there isn't much we can do about that and its also not that critical since its just a single artifact and not many.

Comment (pjfanning) [reply gradle-plugin/build.gradle:58]: you've hardcoded a snapshot url into this publish section so I do not see how this publish can handle when we need to do a non-snapshot release - we are actually waiting to do a release on this very repo

**Review by pjfanning — CHANGES_REQUESTED** [review]
I'm not sure at all about hardcoding snapshot urls into the gradle publishing

File: gradle-plugin/build.gradle (hunk 2 above)
Comment (mdedetrich) [reply gradle-plugin/build.gradle:58]: This is how its documented to be done, from https://github.com/gradle-nexus/publish-plugin#publishing-to-maven-central-via-sonatype-ossrh

> Important. Users registered in Sonatype after [24 February 2021](https://central.sonatype.org/news/20210223_new-users-on-s01/) need to customize the following URLs:
>
 ```
  nexusPublishing {
    repositories {
        sonatype {  //only for users registered in Sonatype after 24 Feb 2021
            nexusUrl.set(uri("https://s01.oss.sonatype.org/service/local/"))
            snapshotRepositoryUrl.set(uri("https://s01.oss.sonatype.org/content/repositories/snapshots/"))
        }
     }
  }
  ```

Comment (pjfanning) [reply gradle-plugin/build.gradle:58]: I understand now. Thanks.

**Review by pjfanning — APPROVED** [review]
lgtm
